### PR TITLE
pytest failure report

### PR DESCRIPTION
**Pull request by xlqian**
A readable failure report will be very helpful!

https://gist.github.com/xlqian/32c4d5b4b0459f9c8a1c2c81c7eb0169

### CODE REVIEW

**Comment by xlqian** [comment]
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_0
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3344%3B48.8686&to=2.395775%3B48.881221999999994&datetime=20190314T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      5,
      {
        "distances":{
          "car":0,
          "walking":5220,
          "bike":0
        },
        "sections":[
          {
            "from":"16 rue Gaillon (Paris)",
            "arrival_date_time":"20190314T201421",
            "departure_date_time":"20190314T190000",
            "to":"78 rue David d'Angers (Paris)",
            "duration":4461,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":4461,
          "total":4461,
          "bike":0
        },
        "arrival_date_time":"20190314T201421",
        "departure_date_time":"20190314T190000",
        "duration":4461,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_4
[query](http://localhost/v1/coverage/idfm/journeys?from=2.2963%3B48.803000000000004&to=2.394357%3B48.833560999999996&datetime=20190314T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 1

<details open><summary>CLICK ME</summary><p>

```json
{
  "3":{
    "$update":{
      "type":"rapid"
    }
  },
  "$insert":[
    [
      4,
      {
        "distances":{
          "car":0,
          "walking":9133,
          "bike":0
        },
        "sections":[
          {
            "from":"77 avenue de la Republique (Ch\u00e2tillon)",
            "arrival_date_time":"20190314T211006",
            "departure_date_time":"20190314T190000",
            "to":"320 rue de Charenton (Paris)",
            "duration":7806,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":7806,
          "total":7806,
          "bike":0
        },
        "arrival_date_time":"20190314T211006",
        "departure_date_time":"20190314T190000",
        "duration":7806,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_5
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3412%3B48.83&to=2.312592%3B48.901918&datetime=20190321T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      7,
      {
        "distances":{
          "car":0,
          "walking":9270,
          "bike":0
        },
        "sections":[
          {
            "from":"93 rue de la Sante (Paris)",
            "arrival_date_time":"20190321T211203",
            "departure_date_time":"20190321T190000",
            "to":"24 rue d'Alsace (Clichy)",
            "duration":7923,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":7923,
          "total":7923,
          "bike":0
        },
        "arrival_date_time":"20190321T211203",
        "departure_date_time":"20190321T190000",
        "duration":7923,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_6
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3521%3B48.8453&to=2.3712299999999997%3B48.877635&datetime=20190307T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      3,
      {
        "distances":{
          "car":0,
          "walking":4283,
          "bike":0
        },
        "sections":[
          {
            "from":"46 rue Monge (Paris)",
            "arrival_date_time":"20190307T200101",
            "departure_date_time":"20190307T190000",
            "to":"120 boulevard de la Villette (Paris)",
            "duration":3661,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":3661,
          "total":3661,
          "bike":0
        },
        "arrival_date_time":"20190307T200101",
        "departure_date_time":"20190307T190000",
        "duration":3661,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_7
[query](http://localhost/v1/coverage/idfm/journeys?from=2.9308%3B49.0362&to=2.343961%3B48.869546&datetime=20190328T180000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      3,
      {
        "distances":{
          "car":0,
          "walking":50418,
          "bike":0
        },
        "sections":[
          {
            "from":"15 rue du Colonel Dubujadoux (\u00c9tr\u00e9pilly)",
            "arrival_date_time":"20190329T055812",
            "departure_date_time":"20190328T180000",
            "to":"33 rue des Jeuneurs (Paris)",
            "duration":43092,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":43092,
          "total":43092,
          "bike":0
        },
        "arrival_date_time":"20190329T055812",
        "departure_date_time":"20190328T180000",
        "duration":43092,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_9
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3576%3B48.873999999999995&to=2.2989919999999997%3B48.849756&datetime=20190328T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      3,
      {
        "distances":{
          "car":0,
          "walking":5951,
          "bike":0
        },
        "sections":[
          {
            "from":"59 boulevard de Magenta (Paris)",
            "arrival_date_time":"20190328T202446",
            "departure_date_time":"20190328T190000",
            "to":"Entr\u00e9e\/Sortie 5|av. de Suffren (Paris)",
            "duration":5086,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":5086,
          "total":5086,
          "bike":0
        },
        "arrival_date_time":"20190328T202446",
        "departure_date_time":"20190328T190000",
        "duration":5086,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_10
[query](http://localhost/v1/coverage/idfm/journeys?from=2.0568%3B48.7847&to=2.297605%3B48.880331&datetime=20190321T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      5,
      {
        "distances":{
          "car":0,
          "walking":23154,
          "bike":0
        },
        "sections":[
          {
            "from":"5 rue Paul Gauguin (Guyancourt)",
            "arrival_date_time":"20190322T002950",
            "departure_date_time":"20190321T190000",
            "to":"27 rue des Renaudes (Paris)",
            "duration":19790,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":19790,
          "total":19790,
          "bike":0
        },
        "arrival_date_time":"20190322T002950",
        "departure_date_time":"20190321T190000",
        "duration":19790,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_11
[query](http://localhost/v1/coverage/idfm/journeys?from=2.2527%3B48.8434&to=2.415858%3B48.6474&datetime=20190328T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 1

<details open><summary>CLICK ME</summary><p>

```json
{
  "10":{
    "$update":{
      "type":"rapid"
    }
  },
  "$insert":[
    [
      11,
      {
        "distances":{
          "car":0,
          "walking":27625,
          "bike":0
        },
        "sections":[
          {
            "from":"14 all\u00e9e Charles Brennus (Paris)",
            "arrival_date_time":"20190329T013331",
            "departure_date_time":"20190328T190000",
            "to":"11 rue du Bearn (Ris-Orangis)",
            "duration":23611,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":23611,
          "total":23611,
          "bike":0
        },
        "arrival_date_time":"20190329T013331",
        "departure_date_time":"20190328T190000",
        "duration":23611,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_12
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3971%3B48.8237&to=2.3466810000000002%3B48.837171999999995&datetime=20190314T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      2,
      {
        "distances":{
          "car":0,
          "walking":4748,
          "bike":0
        },
        "sections":[
          {
            "from":"8 rue du Nouveau Bercy (Charenton-le-Pont)",
            "arrival_date_time":"20190314T200738",
            "departure_date_time":"20190314T190000",
            "to":"67 rue Broca (Paris)",
            "duration":4058,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":4058,
          "total":4058,
          "bike":0
        },
        "arrival_date_time":"20190314T200738",
        "departure_date_time":"20190314T190000",
        "duration":4058,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_14
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3262%3B48.8462&to=2.406677%3B48.87672&datetime=20190328T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      2,
      {
        "distances":{
          "car":0,
          "walking":7380,
          "bike":0
        },
        "sections":[
          {
            "from":"137 rue de Rennes (Paris)",
            "arrival_date_time":"20190328T204507",
            "departure_date_time":"20190328T190000",
            "to":"169 boulevard Mortier (Paris)",
            "duration":6307,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":6307,
          "total":6307,
          "bike":0
        },
        "arrival_date_time":"20190328T204507",
        "departure_date_time":"20190328T190000",
        "duration":6307,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_16
[query](http://localhost/v1/coverage/idfm/journeys?from=2.303%3B48.8992&to=2.383508%3B48.885723999999996&datetime=20190328T190000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      4,
      {
        "distances":{
          "car":0,
          "walking":6508,
          "bike":0
        },
        "sections":[
          {
            "from":"6 rue Casteres (Clichy)",
            "arrival_date_time":"20190328T203243",
            "departure_date_time":"20190328T190000",
            "to":"17 rue de Lorraine (Paris)",
            "duration":5563,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":5563,
          "total":5563,
          "bike":0
        },
        "arrival_date_time":"20190328T203243",
        "departure_date_time":"20190328T190000",
        "duration":5563,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented
## artemis/tests/idfm_test.py::TestIdfMExperimental::test_idfm_17
[query](http://localhost/v1/coverage/idfm/journeys?from=2.3453%3B48.9748&to=2.3658189999999997%3B48.951377&datetime=20190307T210000&datetime_represents=departure&_min_nb_journeys=1&_final_line_filter=True&walking_speed=1.17&_night_bus_filter_max_factor=1.3&_min_journeys_calls=2&_max_extra_second_pass=10) | [open old](http://canaltp.github.io/navitia-playground/file.html)  |  [open new](http://canaltp.github.io/navitia-playground/file.html)
### example_diff:
* example diff not yet implemented
### journeys_diff:
* new journeys nb: 1
* discarded journeys nb: 0
* updated journeys nb: 0

<details open><summary>CLICK ME</summary><p>

```json
{
  "$insert":[
    [
      1,
      {
        "distances":{
          "car":0,
          "walking":3756,
          "bike":0
        },
        "sections":[
          {
            "from":"rue du 11 Novembre 1918 (Montmagny)",
            "arrival_date_time":"20190307T215330",
            "departure_date_time":"20190307T210000",
            "to":"place Jules Verne (Pierrefitte-sur-Seine)",
            "duration":3210,
            "type":"street_network"
          }
        ],
        "nb_transfers":0,
        "durations":{
          "car":0,
          "walking":3210,
          "total":3210,
          "bike":0
        },
        "arrival_date_time":"20190307T215330",
        "departure_date_time":"20190307T210000",
        "duration":3210,
        "type":"non_pt_walk"
      }
    ]
  ]
}
```

</p></details>

### ptref_diff:
* ptref diff not yet implemented

File: artemis/pytest_report_makers.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    with open(ref) as ref_f, open(output) as output_f:
+        ref_j = json.load(ref_f)
+        out_j = json.load(output_f)
+        diff = jsondiff.diff(ref_j['response'].get('journeys', []), out_j['response'].get('journeys', []),
+                             syntax='symmetric')
+
+        updated_nb = sum(str(k).isdigit() for k in diff.keys())
+
+        message = (
+            "* new journeys nb: {}\n"
```
Comment (pbougue) [inline artemis/pytest_report_makers.py]: To improve readability, better to print nothing if nothing changed (at request level as weel as at journey level).
Maybe a quick summary at top level to say "54 short-responses unchanged" or something like that?

File: artemis/conftest.py [hunk 2]
```diff
@@ -83,3 +83,70 @@ def wait_for_cities_completion():
...
+        with open(failures_report, mode) as f:
+            # let's also access a fixture for the fun of it
+            (_, test_class_name, test_name) = rep.nodeid.split("::")
+
+            scenario = retrieve_scenario(test_class_name)
+            test_dataset_name = test_class_name.strip(scenario)
+
+            reference = os.path.join(config['REFERENCE_FILE_PATH'],
+                                     test_dataset_name,
+                                     # yep, that's ugly...
```
Comment (pbougue) [inline artemis/conftest.py]: Maybe mutualize that ugliness so that :hankey: is grouped and not everywhere? :grin:

File: artemis/conftest.py [hunk 3]
```diff
@@ -83,3 +83,70 @@ def wait_for_cities_completion():
...
+                failure_messages.append("\n### {}:".format(fun_name))
+                failure_messages.append(fun(reference, output))
+
+            if not test_dataset_name:
+                log.error(("test {} failed, but data set name cannot be retrieved, "
+                          "no failure report will be generated for this test").format(rep.nodeid))
+
+            f.write("## {}\n".format(rep.nodeid))
+            f.write("[query]({}) | [open old]({})  |  [open new]({})\n".format(query,
+                                                               'http://canaltp.github.io/navitia-playground/file.html',
```
Comment (pbougue) [inline artemis/conftest.py]: Just a TODO in case you forgot: missing file here (or remove that link?).

File: artemis/conftest.py [hunk 4]
```diff
@@ -83,3 +83,70 @@ def wait_for_cities_completion():
...
+
+            # Here is the magic
+            failure_messages = []
+            import inspect
+            for fun_name, fun in inspect.getmembers(pytest_report_makers, inspect.isfunction):
+                failure_messages.append("\n### {}:".format(fun_name))
+                failure_messages.append(fun(reference, output))
+
+            if not test_dataset_name:
+                log.error(("test {} failed, but data set name cannot be retrieved, "
```
Comment (pbougue) [inline artemis/conftest.py]: Maybe a short count in the summary for those could be nice?

**Pull request merged**